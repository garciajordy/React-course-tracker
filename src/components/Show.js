import React, { useEffect, useState } from 'react';
import CircularProgress from '@material-ui/core/CircularProgress';
import axios from 'axios';
import { useHistory } from 'react-router-dom';
import DeleteIcon from '@material-ui/icons/Delete';
import store from '../store/Store';
import AddMeasure from '../actions/AddMeasurementAction';
import RemoveMeasure from '../actions/RemoveMeasurementAction';

function Show() {
  const history = useHistory();
  const storeMeasurements = store.getState().MeasurementReducer;
  const [data, setData] = useState('');
  const [measurements, setMeasurements] = useState([]);
  const course = store.getState().CourseReducer;
  function getMeasurements() {
    axios.get(`http://localhost:3001/courses/${course.id}`, { withCredentials: true }).then((response) => {
      setMeasurements([]);
      setMeasurements(response.data.measurements.map((p) => p));
      measurements.map((data) => {
        const array = storeMeasurements.map((data) => data.id);
        if (array.includes(data.id)) {
          return '';
        }
        return store.dispatch(AddMeasure(data));
      });
    });
  }
  useEffect(() => {
    if (course.name.length < 1) {
      history.push('/dashboard');
    } else {
      getMeasurements();
    }
  }, []);

  function deleteMeasurement(id) {
    axios.delete(`http://localhost:3001/measurements/${id}`, { withCredentials: true })
      .then(() => {
        store.dispatch(RemoveMeasure(id));
        setMeasurements(measurements.filter((item) => parseInt(item.id, 10) !== parseInt(id, 10)));
      })
      .catch((err) => err);
  }
  function handleSubmit(event) {
    axios.post('http://localhost:3001/measurements', {
      measurement: {
        amount: data,
        course_id: course.id,
      },
    }, { withCredentials: true }).then((response) => {
      if (response.data.status === 'created') {
        store.dispatch(AddMeasure(response.data.measurement));
        setMeasurements((prev) => [...prev, response.data.measurement]);
        setData('');
      }
    });
    event.preventDefault();
  }

  function handleChange(e) {
    setData(e.target.value);
  }
  return (
    <div className="show-container">
      <div className="header">
        {course.name}
        .it
      </div>
      <h1 className="dashboard">{course.name}</h1>
      <form onSubmit={handleSubmit}>
        <input type="number" placeholder="Write how many minutes..." value={data} onChange={handleChange} />
        <button className="add-button" type="submit">Add</button>
      </form>
      { measurements.reverse().map((data) => {
        const data2 = new Date(data.created_at.substring(0, 10));
        return (
          <div key={data.id} className="measurement-component">
            <div className="flex-row">

              <CircularProgress variant="determinate" value={data.amount} />
              <p key={data.created_at}>{data2.toString().substring(0, 16)}</p>
            </div>
            <div className="grid-columns">

              <p className="min-para" key={data.amount}>
                {data.amount}
                {' '}
                Min
              </p>
              <button type="button" onClick={() => deleteMeasurement(data.id)} key={data.id}>
                <DeleteIcon />
              </button>
            </div>
          </div>
        );
      })}
    </div>
  );
}

export default Show;
