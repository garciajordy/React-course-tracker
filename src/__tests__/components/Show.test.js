import React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import Show from '../../components/Show';
import store from '../../store/Store';

afterEach(cleanup);

it('should take a snapshot', () => {
  const { asFragment } = render(
    <BrowserRouter>
      <Provider store={store}><Show /></Provider>
    </BrowserRouter>,
  );
  expect(asFragment(
    <BrowserRouter>
      <Provider store={store}><Show /></Provider>
    </BrowserRouter>,
  )).toMatchSnapshot();
});

it('displays heading', () => {
  render(
    <BrowserRouter>
      <Provider store={store}><Show /></Provider>
    </BrowserRouter>,
  );
  expect(screen.getByText('Today')).toBeTruthy();
});
