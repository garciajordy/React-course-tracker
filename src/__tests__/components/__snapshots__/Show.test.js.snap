// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="show-container"
  >
    <div
      class="header"
    >
      .it
    </div>
    <h1
      class="dashboard"
    />
    <form>
      <input
        placeholder="Write how many minutes..."
        type="number"
        value=""
      />
      <button
        class="add-button"
        type="submit"
      >
        Add
      </button>
    </form>
    <div
      class="today"
    >
      <p>
        Today
      </p>
    </div>
  </div>
</DocumentFragment>
`;
